File: android/app/src/main/kotlin/com/example/heart_rate_assessment/HeartRateModule.kt
package com.example.heart_rate_assessment

import android.os.Handler
import android.os.Looper
import android.util.Log
import io.flutter.plugin.common.EventChannel
import java.util.*
import kotlin.random.Random

/**
 * HeartRateModule that simulates connection to an external heart rate monitoring device
 * This module handles the simulated device discovery, connection, and data transmission
 */
class HeartRateModule : EventChannel.StreamHandler {
    private var eventSink: EventChannel.EventSink? = null
    private var handler: Handler? = null
    private val runnable = Runnable { generateAndSendHeartRate() }
    
    // Default heart rate range
    private val normalMinHeartRate = 60
    private val normalMaxHeartRate = 100
    
    // Categories of abnormal heart rates
    private val lowHeartRateRange = 40..59
    private val elevatedHeartRateRange = 101..130
    private val highHeartRateRange = 131..180
    private val criticalHeartRateRange = 181..220
    
    // Configuration parameters
    private val updateIntervalMs = 3000L // 3 seconds
    private val abnormalProbability = 0.2 // 20% chance of abnormal reading
    
    // Simulated device data
    private var connectedDevice: DeviceInfo? = null
    private var deviceStatus = DeviceStatus.DISCONNECTED
    private var connectionQuality = 100 // 0-100%, will fluctuate over time
    
    private val TAG = "HeartRateModule"
    
    /**
     * Set up the event sink when the stream is listened to
     */
    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        Log.d(TAG, "onListen called, stream being listened to")
        eventSink = events
        simulateDeviceDiscovery()
    }
    
    /**
     * Clean up when the stream is cancelled
     */
    override fun onCancel(arguments: Any?) {
        Log.d(TAG, "onCancel called, stream no longer being listened to")
        stopGeneratingHeartRateData()
        simulateDeviceDisconnection()
        eventSink = null
    }
    
    /**
     * Simulate the process of discovering a heart rate monitoring device
     */
    private fun simulateDeviceDiscovery() {
        Log.d(TAG, "Starting device discovery")
        deviceStatus = DeviceStatus.DISCOVERING
        
        // Generate a synthetic device if none exists
        if (connectedDevice == null) {
            createRandomDevice()
        }
        
        // Send device discovery event
        sendDeviceStatusUpdate("Device found: ${connectedDevice?.name}")
        
        // After a short delay, connect to the device
        Handler(Looper.getMainLooper()).postDelayed({
            simulateDeviceConnection()
        }, 1500) // 1.5 second delay to simulate discovery time
    }
    
    /**
     * Create a random heart rate monitoring device
     */
    private fun createRandomDevice() {
        val random = Random
        
        // Select random device type
        val deviceModels = listOf(
            DeviceInfo("BeatMaster Pro", "CardioTech", "3.1.4", "Chest Strap"),
            DeviceInfo("HeartSense Ultra", "FitLife", "2.8.0", "Wristband"),
            DeviceInfo("PulseTrack Elite", "SportMedix", "4.2.1", "Chest Strap"),
            DeviceInfo("CardioRhythm X2", "HealthSystems", "5.0.3", "Medical Monitor"),
            DeviceInfo("VitalPulse Watch", "WearTech", "1.9.7", "Smartwatch")
        )
        
        val selectedDevice = deviceModels[random.nextInt(deviceModels.size)]
        
        // Create a unique device ID
        val deviceId = "HR-${random.nextInt(999999).toString().padStart(6, '0')}"
        
        // Random battery level (20-100%)
        val batteryLevel = 20 + random.nextInt(81)
        
        // Random signal strength (60-95%)
        val signalStrength = 60 + random.nextInt(36)
        
        connectedDevice = selectedDevice.copy(
            id = deviceId,
            batteryLevel = batteryLevel,
            signalStrength = signalStrength
        )
        
        Log.d(TAG, "Created random device: ${connectedDevice?.name} (${connectedDevice?.id})")
    }
    
    /**
     * Simulate connecting to the heart rate device
     */
    private fun simulateDeviceConnection() {
        if (connectedDevice == null) {
            Log.e(TAG, "No device available for connection")
            sendDeviceStatusUpdate("Error: No device found")
            deviceStatus = DeviceStatus.ERROR
            return
        }
        
        Log.d(TAG, "Connecting to device: ${connectedDevice?.name}")
        deviceStatus = DeviceStatus.CONNECTING
        sendDeviceStatusUpdate("Connecting to ${connectedDevice?.name}...")
        
        // Simulate connection process with a delay
        Handler(Looper.getMainLooper()).postDelayed({
            // 90% chance of successful connection
            if (Random.nextDouble() < 0.9) {
                deviceStatus = DeviceStatus.CONNECTED
                Log.d(TAG, "Successfully connected to ${connectedDevice?.name}")
                sendDeviceStatusUpdate("Connected to ${connectedDevice?.name}")
                
                // Start sending heart rate data
                startGeneratingHeartRateData()
            } else {
                // Simulate a connection failure
                deviceStatus = DeviceStatus.ERROR
                Log.e(TAG, "Failed to connect to ${connectedDevice?.name}")
                sendDeviceStatusUpdate("Error: Connection failed")
            }
        }, 2000) // 2 second delay to simulate connection time
    }
    
    /**
     * Simulate disconnecting from the heart rate device
     */
    private fun simulateDeviceDisconnection() {
        if (deviceStatus == DeviceStatus.CONNECTED || deviceStatus == DeviceStatus.MONITORING) {
            Log.d(TAG, "Disconnecting from device: ${connectedDevice?.name}")
            deviceStatus = DeviceStatus.DISCONNECTING
            sendDeviceStatusUpdate("Disconnecting from ${connectedDevice?.name}...")
            
            // Simulate disconnection process
            Handler(Looper.getMainLooper()).postDelayed({
                deviceStatus = DeviceStatus.DISCONNECTED
                Log.d(TAG, "Disconnected from ${connectedDevice?.name}")
                sendDeviceStatusUpdate("Disconnected from ${connectedDevice?.name}")
            }, 1000)
        }
    }
    
    /**
     * Send device status update to Flutter
     */
    private fun sendDeviceStatusUpdate(message: String) {
        val statusData = HashMap<String, Any>()
        statusData["type"] = "deviceStatus"
        statusData["message"] = message
        statusData["status"] = deviceStatus.name
        
        connectedDevice?.let {
            statusData["device"] = mapOf(
                "id" to it.id,
                "name" to it.name,
                "manufacturer" to it.manufacturer,
                "firmwareVersion" to it.firmwareVersion,
                "type" to it.type,
                "batteryLevel" to it.batteryLevel,
                "signalStrength" to it.signalStrength
            )
        }
        
        eventSink?.let {
            Log.d(TAG, "Sending device status update to Flutter")
            it.success(statusData)
        }
    }
    
    /**
     * Start generating random heart rate data every 5 seconds
     */
    fun startGeneratingHeartRateData() {
        Log.d(TAG, "startGeneratingHeartRateData called")
        deviceStatus = DeviceStatus.MONITORING
        stopGeneratingHeartRateData() // Ensure no duplicate timers
        
        handler = Handler(Looper.getMainLooper())
        handler?.postDelayed(runnable, updateIntervalMs)
        
        // Generate and send first value immediately
        generateAndSendHeartRate()
    }
    
    /**
     * Stop generating heart rate data
     */
    fun stopGeneratingHeartRateData() {
        Log.d(TAG, "stopGeneratingHeartRateData called")
        handler?.removeCallbacks(runnable)
        handler = null
    }
    
    /**
     * Generate a random heart rate value and send via eventSink
     */
    private fun generateAndSendHeartRate() {
        if (deviceStatus != DeviceStatus.MONITORING || connectedDevice == null) {
            Log.d(TAG, "Not in monitoring state, won't generate heart rate")
            return
        }
        
        // Update connection quality (will fluctuate slightly)
        updateConnectionQuality()
        
        // Check if we should simulate a connection issue
        if (connectionQuality < 30) {
            simulateConnectionIssue()
            return
        }
        
        // Get the adjusted heart rate based on device accuracy and connection quality
        val heartRate = generateHeartRateValue()
        val timestamp = System.currentTimeMillis() / 1000.0
        
        // Create a map with heart rate value, timestamp, and device info
        val heartRateData = HashMap<String, Any>()
        heartRateData["type"] = "heartRate"
        heartRateData["heartRate"] = heartRate
        heartRateData["timestamp"] = timestamp
        heartRateData["connectionQuality"] = connectionQuality
        
        // Include device information
        connectedDevice?.let {
            heartRateData["device"] = mapOf(
                "id" to it.id,
                "name" to it.name,
                "manufacturer" to it.manufacturer,
                "type" to it.type,
                "batteryLevel" to it.batteryLevel,
                "signalStrength" to it.signalStrength
            )
        }
        
        Log.d(TAG, "Generating heart rate: $heartRate BPM, quality: $connectionQuality%")
        
        // Send the heart rate data to Flutter
        eventSink?.let {
            Log.d(TAG, "Sending heart rate data to Flutter")
            it.success(heartRateData)
        } ?: run {
            Log.e(TAG, "EventSink is null, cannot send heart rate data")
        }
        
        // Schedule the next update
        handler?.postDelayed(runnable, updateIntervalMs)
    }
    
    /**
     * Update the connection quality to simulate real-world fluctuations
     */
    private fun updateConnectionQuality() {
        val random = Random
        
        // Simulate signal fluctuation
        val fluctuation = random.nextInt(11) - 5 // -5 to +5 change
        connectionQuality += fluctuation
        
        // Ensure quality stays within bounds
        connectionQuality = connectionQuality.coerceIn(0, 100)
        
        // Update device signal strength occasionally
        if (random.nextInt(10) == 0) { // 10% chance
            connectedDevice?.let {
                val newSignalStrength = it.signalStrength + random.nextInt(11) - 5
                connectedDevice = it.copy(signalStrength = newSignalStrength.coerceIn(0, 100))
            }
        }
        
        // Update device battery level occasionally (should decrease slowly)
        if (random.nextInt(20) == 0) { // 5% chance
            connectedDevice?.let {
                val newBatteryLevel = it.batteryLevel - 1
                connectedDevice = it.copy(batteryLevel = newBatteryLevel.coerceAtLeast(0))
                
                // If battery gets low, send a warning
                if (newBatteryLevel <= 20 && newBatteryLevel % 5 == 0) {
                    sendDeviceStatusUpdate("Warning: ${it.name} battery low (${newBatteryLevel}%)")
                }
            }
        }
    }
    
    /**
     * Simulate a connection issue with the device
     */
    private fun simulateConnectionIssue() {
        Log.d(TAG, "Simulating connection issue, quality: $connectionQuality%")
        
        // Create status data
        val statusData = HashMap<String, Any>()
        statusData["type"] = "connectionIssue"
        statusData["message"] = "Weak signal from ${connectedDevice?.name}"
        statusData["connectionQuality"] = connectionQuality
        
        // Send to Flutter
        eventSink?.success(statusData)
        
        // We'll still try to schedule the next update
        handler?.postDelayed(runnable, updateIntervalMs)
    }
    
    /**
     * Generate a heart rate value with potential for abnormal readings
     * Adjusts based on device accuracy and connection quality
     */
    private fun generateHeartRateValue(): Int {
        val baseRate = if (Random.nextDouble() < abnormalProbability) {
            // Generate an abnormal reading
            when (Random.nextInt(4)) {
                0 -> lowHeartRateRange.random() // Low heart rate
                1 -> elevatedHeartRateRange.random() // Elevated heart rate
                2 -> highHeartRateRange.random() // High heart rate
                else -> criticalHeartRateRange.random() // Critical heart rate
            }.also {
                Log.d(TAG, "Generated abnormal heart rate: $it BPM")
            }
        } else {
            // Generate a normal heart rate
            Random.nextInt(normalMinHeartRate, normalMaxHeartRate + 1).also {
                Log.d(TAG, "Generated normal heart rate: $it BPM")
            }
        }
        
        // If connection quality or device accuracy is low, add some noise to the reading
        val deviceAccuracy = connectedDevice?.accuracy ?: 100
        val noiseLevel = (100 - ((connectionQuality + deviceAccuracy) / 2)) / 10
        
        // Add random noise based on the calculated noise level
        val noise = if (noiseLevel > 0) Random.nextInt(-noiseLevel, noiseLevel + 1) else 0
        
        return (baseRate + noise).coerceIn(30, 230) // Keep within physiological limits
    }
    
    /**
     * Manually trigger an error for testing
     */
    fun sendError() {
        Log.e(TAG, "Manually sending error for testing")
        eventSink?.error(
            "HEART_RATE_ERROR",
            "Failed to generate heart rate data",
            null
        )
    }
    
    /**
     * Data class to store information about a heart rate monitoring device
     */
    data class DeviceInfo(
        val name: String,
        val manufacturer: String,
        val firmwareVersion: String,
        val type: String,
        val id: String = "",
        val batteryLevel: Int = 100,
        val signalStrength: Int = 100,
        val accuracy: Int = 95
    )
    
    /**
     * Enum describing the possible states of the device connection
     */
    enum class DeviceStatus {
        DISCONNECTED,   // No device connected
        DISCOVERING,    // Searching for devices
        CONNECTING,     // Attempting to connect
        CONNECTED,      // Connected but not yet monitoring
        MONITORING,     // Connected and actively monitoring
        DISCONNECTING,  // In the process of disconnecting
        ERROR           // An error occurred
    }
} 

File: ios/Runner/HeartRateModule.swift
import Foundation
import Flutter
import os.log

/// HeartRateModule that generates random heart rate values and simulates device connectivity
class HeartRateModule: NSObject, FlutterStreamHandler {
    // Flutter channels
    private let methodChannel: FlutterMethodChannel
    private let eventChannel: FlutterEventChannel
    
    // Flutter event sink
    private var eventSink: FlutterEventSink?
    private var timer: Timer?
    
    // Configuration parameters
    private let updateIntervalSec = 1.0 // 1 second
    
    // Heart rate ranges
    private let normalRange = 60...100
    private let lowRange = 40..<60
    private let elevatedRange = 101...120
    private let highRange = 121...140
    private let criticalRange = 141...180
    
    // Current simulation state
    private var currentHeartRate = 75
    private var heartRateCategory = "NORMAL"
    private var abnormalReadingProbability = 0.05
    
    // Simulated device
    private var device = DeviceInfo.random()
    private var deviceUpdateCounter = 0
    
    private let logger = OSLog(subsystem: "com.example.heart_rate_assessment", category: "HeartRateModule")
    
    init(messenger: FlutterBinaryMessenger) {
        // Initialize channels
        methodChannel = FlutterMethodChannel(
            name: "com.example.heart_rate_assessment/heart_rate_method",
            binaryMessenger: messenger
        )
        
        eventChannel = FlutterEventChannel(
            name: "com.example.heart_rate_assessment/heart_rate_stream",
            binaryMessenger: messenger
        )
        
        super.init()
        
        // Set up method channel handlers
        setupMethodChannel()
        
        // Set up event channel
        eventChannel.setStreamHandler(self)
        
        os_log("HeartRateModule initialized with device: %@", log: logger, type: .info, device.name)
    }
    
    private func setupMethodChannel() {
        methodChannel.setMethodCallHandler { [weak self] (call, result) in
            guard let self = self else { return }
            
            switch call.method {
            case "startHeartRateMonitoring":
                self.startHeartRateMonitoring()
                result(true)
            case "stopHeartRateMonitoring":
                self.stopHeartRateMonitoring()
                result(true)
            default:
                result(FlutterMethodNotImplemented)
            }
        }
    }
    
    // MARK: - FlutterStreamHandler protocol
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        os_log("Stream is being listened to", log: logger, type: .debug)
        eventSink = events
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        os_log("Stream is no longer being listened to", log: logger, type: .debug)
        stopHeartRateMonitoring()
        eventSink = nil
        return nil
    }
    
    // MARK: - Public methods
    
    func startHeartRateMonitoring() {
        os_log("Starting heart rate monitoring", log: logger, type: .debug)
        stopHeartRateMonitoring() // Ensure no duplicate timers
        
        // Reset device if needed
        device = DeviceInfo.random()
        
        // Generate and send first value immediately
        generateAndSendHeartRate()
        
        // Schedule periodic updates
        timer = Timer.scheduledTimer(
            timeInterval: updateIntervalSec,
            target: self,
            selector: #selector(timerFired),
            userInfo: nil,
            repeats: true
        )
    }
    
    func stopHeartRateMonitoring() {
        os_log("Stopping heart rate monitoring", log: logger, type: .debug)
        timer?.invalidate()
        timer = nil
    }
    
    // MARK: - Private methods
    
    @objc private func timerFired() {
        generateAndSendHeartRate()
    }
    
    private func generateAndSendHeartRate() {
        guard let sink = eventSink else {
            os_log("EventSink is nil, cannot send heart rate data", log: logger, type: .error)
            return
        }
        
        // Update device status occasionally
        deviceUpdateCounter += 1
        if deviceUpdateCounter >= 5 {
            updateDeviceStatus()
            deviceUpdateCounter = 0
        }
        
        // Generate a heart rate value
        let useAbnormalReading = Double.random(in: 0...1) < abnormalReadingProbability
        
        if useAbnormalReading {
            // Generate an abnormal reading occasionally
            let abnormalCategories = ["LOW", "ELEVATED", "HIGH", "CRITICAL"]
            heartRateCategory = abnormalCategories.randomElement() ?? "ELEVATED"
            
            switch heartRateCategory {
            case "LOW":
                currentHeartRate = Int.random(in: lowRange)
            case "ELEVATED":
                currentHeartRate = Int.random(in: elevatedRange)
            case "HIGH":
                currentHeartRate = Int.random(in: highRange)
            case "CRITICAL":
                currentHeartRate = Int.random(in: criticalRange)
            default:
                currentHeartRate = Int.random(in: normalRange)
            }
        } else {
            // Most of the time, generate values within normal range
            heartRateCategory = "NORMAL"
            
            // Small random fluctuation around the current value
            let fluctuation = Int.random(in: -5...5)
            currentHeartRate += fluctuation
            
            // Ensure it stays within normal range
            if currentHeartRate < normalRange.lowerBound {
                currentHeartRate = normalRange.lowerBound + Int.random(in: 0...5)
            } else if currentHeartRate > normalRange.upperBound {
                currentHeartRate = normalRange.upperBound - Int.random(in: 0...5)
            }
        }
        
        // Create the heart rate data
        let timestamp = Date().timeIntervalSince1970 * 1000
        
        var heartRateData: [String: Any] = [
            "heartRate": currentHeartRate,
            "timestamp": timestamp
        ]
        
        // Add device info
        heartRateData["device"] = device.toDictionary()
        
        os_log("Sending heart rate: %d BPM", log: logger, type: .debug, currentHeartRate)
        sink(heartRateData)
    }
    
    private func updateDeviceStatus() {
        // Update battery level - slowly decrease
        if device.batteryLevel > 10 {
            device.batteryLevel -= Int.random(in: 1...3)
            if device.batteryLevel < 0 {
                device.batteryLevel = 0
            }
        }
        
        // Update signal strength - fluctuate
        let signalChange = Int.random(in: -1...1)
        device.signalStrength += signalChange
        if device.signalStrength < 1 {
            device.signalStrength = 1
        } else if device.signalStrength > 5 {
            device.signalStrength = 5
        }
        
        // Simulate occasional connection issues
        if Double.random(in: 0...1) < 0.05 {
            device.connectionStatus = ["CONNECTED", "CONNECTING", "UNSTABLE"].randomElement() ?? "CONNECTED"
        } else {
            device.connectionStatus = "CONNECTED"
        }
    }
}

// MARK: - Device Info Model
struct DeviceInfo {
    var id: String
    var name: String
    var type: String
    var manufacturer: String
    var firmwareVersion: String
    var batteryLevel: Int
    var signalStrength: Int
    var connectionStatus: String
    var accuracy: String
    
    func toDictionary() -> [String: Any] {
        return [
            "id": id,
            "name": name,
            "type": type,
            "manufacturer": manufacturer,
            "firmwareVersion": firmwareVersion,
            "batteryLevel": batteryLevel,
            "signalStrength": signalStrength,
            "connectionStatus": connectionStatus,
            "accuracy": accuracy,
            "signalDescription": getSignalDescription()
        ]
    }
    
    func getSignalDescription() -> String {
        switch signalStrength {
        case 1:
            return "Very Weak"
        case 2:
            return "Weak"
        case 3:
            return "Moderate"
        case 4:
            return "Good"
        case 5:
            return "Excellent"
        default:
            return "Unknown"
        }
    }
    
    static func random() -> DeviceInfo {
        let deviceNames = ["HeartSense Pro", "CardioMonitor X2", "BeatMaster Plus", "PulseTrack Elite", "VitalScan Ultra"]
        let deviceTypes = ["PPG Sensor", "ECG Monitor", "Dual Mode Sensor", "Multi-Parameter Monitor"]
        let manufacturers = ["HealthTech", "CardioInnovations", "MedSense", "VitalMetrics", "BioTech Solutions"]
        let accuracyLevels = ["High", "Medical Grade", "Consumer", "Research Grade", "Clinical"]
        
        return DeviceInfo(
            id: UUID().uuidString,
            name: deviceNames.randomElement() ?? "HeartSense Pro",
            type: deviceTypes.randomElement() ?? "PPG Sensor",
            manufacturer: manufacturers.randomElement() ?? "HealthTech",
            firmwareVersion: "v\(Int.random(in: 1...5)).\(Int.random(in: 0...9)).\(Int.random(in: 0...9))",
            batteryLevel: Int.random(in: 50...100),
            signalStrength: Int.random(in: 3...5),
            connectionStatus: "CONNECTED",
            accuracy: accuracyLevels.randomElement() ?? "High"
        )
    }
} 